File: android/src/ti/airship/TIAirshipModule.java
/* Copyright Airship and Contributors */

package ti.airship;

import com.urbanairship.Autopilot;
import com.urbanairship.UAirship;
import com.urbanairship.actions.ActionRunRequest;
import com.urbanairship.actions.AddCustomEventAction;
import com.urbanairship.automation.InAppAutomation;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;
import com.urbanairship.json.JsonValue;
import com.urbanairship.messagecenter.MessageCenter;
import com.urbanairship.util.UAStringUtil;

import org.appcelerator.kroll.KrollFunction;
import org.appcelerator.kroll.KrollModule;
import org.appcelerator.kroll.KrollProxy;
import org.appcelerator.kroll.annotations.Kroll;
import org.appcelerator.kroll.common.Log;
import org.appcelerator.titanium.TiApplication;

import java.util.HashMap;
import java.util.HashSet;

import ti.airship.events.EventEmitter;

@Kroll.module(name = "ti.airship", id = "ti.airship")
public class TIAirshipModule extends KrollModule {

    @Kroll.constant
    public static final String EVENT_CHANNEL_UPDATED = "EVENT_CHANNEL_UPDATED";

    @Kroll.constant
    public static final String EVENT_DEEP_LINK_RECEIVED = "EVENT_DEEP_LINK_RECEIVED";

    @Kroll.constant
    public static final String EVENT_PUSH_RECEIVED = "EVENT_PUSH_RECEIVED";

    @Kroll.constant
    public static final String EVENT_NOTIFICATION_RESPONSE = "EVENT_NOTIFICATION_RESPONSE";

    @Kroll.constant
    public static final String EVENT_NOTIFICATION_OPT_IN_CHANGED = "EVENT_NOTIFICATION_OPT_IN_CHANGED";

    private static final String MODULE_NAME = "AirshipTitanium";

    public TIAirshipModule() {
        super(MODULE_NAME);
    }

    @Kroll.onAppCreate
    public static void onAppCreate(TiApplication app) {
        Autopilot.automaticTakeOff(app);
    }

    @Override
    public void listenerAdded(String type, int count, KrollProxy proxy) {
        super.listenerAdded(type, count, proxy);
        EventEmitter.shared().addListeners(type, count, this);
    }

    @Override
    public void listenerRemoved(String type, int count, KrollProxy proxy) {
        super.listenerRemoved(type, count, proxy);
        EventEmitter.shared().removeListeners(type, count, this);
    }

    @Kroll.method
    public void displayMessageCenter() {
        MessageCenter.shared().showMessageCenter();
    }


}

File: android/src/ti/airship/TiAirshipModule.kt
/* Copyright Airship and Contributors */
package ti.airship

import androidx.arch.core.util.Function
import org.appcelerator.kroll.annotations.Kroll.module
import org.appcelerator.kroll.KrollModule
import ti.airship.components.ChannelProxy
import ti.airship.components.AnalyticsProxy
import ti.airship.components.ContactProxy
import ti.airship.components.InAppAutomationProxy
import ti.airship.components.LocaleProxy
import ti.airship.components.MessageCenterProxy
import ti.airship.components.PreferenceCenterProxy
import ti.airship.components.PrivacyManagerProxy
import ti.airship.components.PushProxy
import org.appcelerator.kroll.annotations.Kroll.method
import org.appcelerator.kroll.annotations.Kroll.getProperty
import com.urbanairship.UAirship
import org.appcelerator.kroll.KrollDict
import ti.airship.utils.PluginLogger
import com.urbanairship.json.JsonValue
import ti.airship.utils.PluginStore
import com.urbanairship.Autopilot
import org.appcelerator.titanium.TiApplication
import org.appcelerator.kroll.KrollProxy
import org.appcelerator.kroll.annotations.Kroll.constant
import ti.airship.events.Event
import ti.airship.events.EventManager
import ti.airship.utils.ConfigParser

@module(name = "TiAirship", id = "ti.airship")
class TiAirshipModule : KrollModule() {

    companion object {
        @constant
        val siteEu = "eu"
        @constant
        val siteUs = "US"

        @constant
        val logLevelVerbose = "verbose"
        @constant
        val logLevelDebug = "debug"
        @constant
        val logLevelInfo = "info"
        @constant
        val logLevelWarning = "warning"
        @constant
        val logLevelError = "error"
        @constant
        val logLevelNone = "none"

        @constant
        val eventPushReceived = "push_received"
        @constant
        val eventNotificationResponseReceived = "notification_response_received"
        @constant
        val eventNotificationOptInStatusChanged = "notification_opt_in_status_changed"
        @constant
        val eventChannelCreated = "channel_created"
        @constant
        val eventDeepLinkReceived = "deep_link_received"
        @constant
        val eventInboxUpdated = "inbox_updated"
        @constant
        val eventOpenPreferenceCenter = "open_preference_Center"
    }

    @get:getProperty
    @get:method
    val channel = ChannelProxy()

    @get:getProperty
    @get:method
    val analytics = AnalyticsProxy()

    @get:getProperty
    @get:method
    val contact = ContactProxy()

    @get:getProperty
    @get:method
    val inAppAutomation = InAppAutomationProxy()

    @get:getProperty
    @get:method
    val locale = LocaleProxy()

    @get:getProperty
    @get:method
    val messageCenter = MessageCenterProxy()

    @get:getProperty
    @get:method
    val preferenceCenter = PreferenceCenterProxy()

    @get:getProperty
    @get:method
    val privacyManager = PrivacyManagerProxy()

    @get:getProperty
    @get:method
    val push = PushProxy()

    @get:getProperty
    @get:method
    val isFlying: Boolean
        get() = UAirship.isFlying() || UAirship.isTakingOff()

    @method
    fun takeOff(config: KrollDict?): Boolean {
        PluginLogger.logCall("takeOff", config)
        checkNotNull(config)

        val map = JsonValue.wrapOpt(config).map
        ConfigParser.parse(getActivity(), map)?.validate()

        PluginStore.shared(getActivity()).airshipConfig = map
        Autopilot.automaticTakeOff(TiApplication.getInstance())
        return UAirship.isFlying() || UAirship.isTakingOff()
    }

    init {
        EventManager.setDispatcher((Function { input: Event ->
            fireEvent(
                input.name,
                input.data
            )
        } as Function<Event, Boolean>))
    }

    override fun listenerAdded(type: String, count: Int, proxy: KrollProxy) {
        super.listenerAdded(type, count, proxy)
        EventManager.onListenerAdded()
    }
}